org/mozilla/jss/ssl/javax/JSSSocketChannel.java

```diff
@@ -0,0 +1,465 @@
+package org.mozilla.jss.ssl.javax;
+
+import java.io.*;
+import java.net.*;
+import java.nio.ByteBuffer;
+import java.nio.channels.*;
+import java.security.*;
+import java.util.*;
+
+import javax.net.ssl.*;
+
+/**
+ * SSL-enabled SocketChannel following the javax.net.ssl.SSLSocket interface.
+ *
+ * This class should never be constructed directly and instead only accessed
+ * once a Socket is wrapped in a JSSSocket.
+ *
+ * This class contains all low-level interactions with the underlying
+ * SSLEngine and reading/writing to/from the underlying Socket.
+ */
+public class JSSSocketChannel extends SocketChannel {
+    private JSSSocket sslSocket;
+    private SocketChannel parent;
+    private Socket parentSocket;
+    private ReadableByteChannel readChannel;
+    private WritableByteChannel writeChannel;
+    private JSSEngine engine;
+
+    private InputStream consumed;
+    private boolean autoClose = true;
+
+    private boolean inboundClosed = false;
+    private boolean outboundClosed = false;
+
```

Comment:
Move byte buffers here and make them sized based on `engine.getSession().getApplicationBufferSize()` rather than the internal constant value.

Reply:
Could also move `empty` here.